### PR TITLE
Rewrite image to TypeScript

### LINKED ISSUE #13017: [TS] Rewrite ckeditor5-image into TypeScript
Part of #11704.

### PR DESCRIPTION
**Pull request by mremiszewski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (image): Rewrite ckeditor-image into Typescript. Closes #13017.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-image/src/image/utils.ts [hunk 1]
```diff
@@ -29,30 +39,28 @@ export function createInlineImageViewElement( writer ) {
...
  * @protected
- * @param {module:engine/view/downcastwriter~DowncastWriter} writer
- * @returns {module:engine/view/containerelement~ContainerElement}
  */
-export function createBlockImageViewElement( writer ) {
+export function createBlockImageViewElement( writer: DowncastWriter ): ViewContainerElement {
 	return writer.createContainerElement( 'figure', { class: 'image' }, [
 		writer.createEmptyElement( 'img' ),
-		writer.createSlot()
+		writer.createSlot( undefined as any )
```
Comment (niegowski) [inline packages/ckeditor5-image/src/image/utils.ts:53]: `writer.createSlot()` should have a default parameter.

File: packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts [hunk 2]
```diff
@@ -22,45 +22,45 @@ import ImageCaptionUtils from './imagecaptionutils';
...
 		return 'ImageCaptionEditing';
 	}
 
+	/**
+	 * A map that keeps saved JSONified image captions and image model elements they are
+	 * associated with.
+	 *
+	 * To learn more about this system, see {@link #_saveCaption}.
+	 *
+	 * @member {WeakMap.<module:engine/model/element~Element,Object>}
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts:47]: This is not needed.

File: packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts [hunk 3]
```diff
@@ -90,7 +90,7 @@ export default class ImageCaptionEditing extends Plugin {
 	 *
 	 * @private
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts:91]: This line is not needed.

File: packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts [hunk 4]
```diff
@@ -263,11 +258,11 @@ export default class ImageCaptionEditing extends Plugin {
 			const changes = model.document.differ.getChanges();
 
 			for ( const change of changes ) {
-				if ( change.attributeKey !== 'alt' ) {
+				if ( ( change as any ).attributeKey !== 'alt' ) {
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imagecaption/imagecaptionediting.ts:261]: Please avoid `any`.

File: packages/ckeditor5-image/src/imagecaption/imagecaptionutils.ts [hunk 5]
```diff
@@ -37,7 +36,7 @@ export default class ImageCaptionUtils extends Plugin {
 	 * @param {module:engine/model/element~Element} imageModelElement
 	 * @returns {module:engine/model/element~Element|null}
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imagecaption/imagecaptionutils.ts:37]: Please remove JSDoc types.

Comment (niegowski) [reply packages/ckeditor5-image/src/imagecaption/imagecaptionutils.ts:37]: Here and in other places.

File: packages/ckeditor5-image/src/imagecaption/imagecaptionutils.ts [hunk 6]
```diff
@@ -76,14 +75,20 @@ export default class ImageCaptionUtils extends Plugin {
 	 * @returns {Object|null} Returns the object accepted by {@link module:engine/view/matcher~Matcher} or `null` if the element
 	 * cannot be matched.
 	 */
-	matchImageCaptionViewElement( element ) {
+	public matchImageCaptionViewElement( element: ViewElement ): object | null {
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imagecaption/imagecaptionutils.ts:78]: We should not use `object`.

**Review by niegowski — CHANGES_REQUESTED** [review]
Partial review.

File: packages/ckeditor5-image/src/imageinsert/imageinsertui.ts [hunk 7]
```diff
@@ -20,23 +21,28 @@ import { prepareIntegrations } from './utils';
...
 	 */
-	static get pluginName() {
+	public static get pluginName(): 'ImageInsertUI' {
 		return 'ImageInsertUI';
 	}
 
+	/**
+	 * The dropdown view responsible for displaying the image insert UI.
+	 *
+	 * @member {module:ui/dropdown/dropdownview~DropdownView}
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/imageinsertui.ts:36]:
```suggestion
```

File: packages/ckeditor5-image/src/imageinsert/imageinsertui.ts [hunk 8]
```diff
@@ -83,7 +81,7 @@ export default class ImageInsertUI extends Plugin {
 		} );
 
 		if ( uploadImageCommand ) {
-			const splitButtonView = this.dropdownView.buttonView;
+			const splitButtonView = this.dropdownView.buttonView as any;
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/imageinsertui.ts:84]: Please avoid `any`.

File: packages/ckeditor5-image/src/imageinsert/ui/imageinsertformrowview.ts [hunk 9]
```diff
@@ -19,70 +20,72 @@ import '../../../theme/imageinsertformrowview.css';
...
+	 * An additional CSS class added to the {@link #element}.
+	 *
+	 * @observable
+	 */
+	declare public class: string | null;
+
+	/**
+	 * A collection of row items (buttons, dropdowns, etc.).
+	 *
+	 * @readonly
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/ui/imageinsertformrowview.ts:35]: This is not needed.

File: packages/ckeditor5-image/src/imageinsert/ui/imageinsertformrowview.ts [hunk 10]
```diff
@@ -19,70 +20,72 @@ import '../../../theme/imageinsertformrowview.css';
...
+	 */
+	declare public _role: string | null;
+
+	/**
+	 * The ARIA property reflected by the `aria-labelledby` DOM attribute of the {@link #element}.
+	 *
+	 * **Note**: Used only when a `labelView` is passed to constructor `options`.
+	 *
+	 * @observable
+	 * @privtae
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/ui/imageinsertformrowview.ts:55]:
```suggestion
	 * @private
```

File: packages/ckeditor5-image/src/imageinsert/ui/imageinsertpanelview.ts [hunk 11]
```diff
@@ -102,28 +102,20 @@ export default class ImageInsertPanelView extends View {
...
-		this.set( '_integrations', new Collection() );
+		this.set( '_integrations', new Collection<View>() );
 
-		if ( integrations ) {
-			for ( const [ integration, integrationView ] of Object.entries( integrations ) ) {
-				if ( integration === 'insertImageViaUrl' ) {
-					integrationView.fieldView.bind( 'value' ).to( this, 'imageURLInputValue', value => value || '' );
+		for ( const [ integration, integrationView ] of Object.entries( integrations ) ) {
+			if ( integration === 'insertImageViaUrl' ) {
+				( integrationView as any ).fieldView.bind( 'value' ).to( this, 'imageURLInputValue', ( value: string ) => value || '' );
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/ui/imageinsertpanelview.ts:109]: This probably should be some common interface. Please avoid `any`.

File: packages/ckeditor5-image/src/imageinsert/ui/imageinsertpanelview.ts [hunk 12]
```diff
@@ -209,11 +201,10 @@ export default class ImageInsertPanelView extends View {
...
 	 * Returns a view of the integration.
 	 *
-	 * @param {String} name The name of the integration.
-	 * @returns {module:ui/view~View}
+	 * @param name The name of the integration.
 	 */
-	getIntegration( name ) {
-		return this._integrations.find( integration => integration.name === name );
+	public getIntegration( name: string ): View {
+		return this._integrations.find( ( integration: any ) => integration.name === name )!;
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/ui/imageinsertpanelview.ts:207]: Please avoid `any`.

File: packages/ckeditor5-image/src/imageinsert/utils.ts [hunk 13]
```diff
@@ -17,11 +19,11 @@ import { LabeledFieldView, createLabeledInputText } from 'ckeditor5/src/ui';
  *
  * @returns {Object.<String, module:ui/view~View>} Integrations object.
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/utils.ts:20]: Remember to strip types in JSDoc comments. I can see plenty of those left in many files. Please find them all and remove.

File: packages/ckeditor5-image/src/imageinsert/utils.ts [hunk 14]
```diff
@@ -17,11 +19,11 @@ import { LabeledFieldView, createLabeledInputText } from 'ckeditor5/src/ui';
...
  */
-export function prepareIntegrations( editor ) {
+export function prepareIntegrations( editor: Editor ): Record<string, View> {
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageinsert/utils.ts:22]: Maybe this should return sth more specific than `View` ? some dedicated interface?

**Review by niegowski — CHANGES_REQUESTED** [review]
Partial review. Please remove all JSDoc types and not needed doc tags (for example `@member`, `@extends` etc). Also avoid using `as any`.

File: packages/ckeditor5-image/src/imageinsert/imageinsertui.ts (hunk 8 above)
Comment (mremiszewski) [reply packages/ckeditor5-image/src/imageinsert/imageinsertui.ts:84]: I can't eliminate `as any` here. There's an assignment to `actionView` which is `readonly` field.

File: packages/ckeditor5-image/src/imageinsert/utils.ts (hunk 14 above)
Comment (mremiszewski) [reply packages/ckeditor5-image/src/imageinsert/utils.ts:22]: I'm not sure if that can be done. In tests, some integrations are defined as `View`

**Review by arkflpc — COMMENTED** [review]
Partial review done. To be continued...

File: packages/ckeditor5-image/src/imageutils.ts [hunk 15]
```diff
@@ -204,115 +202,105 @@ export default class ImageUtils extends Plugin {
...
-//
-// @private
-// @param {module:engine/model/selection~Selection} selection
-// @param {module:engine/model/model~Model} model
-// @returns {module:engine/model/element~Element}
-function getInsertImageParent( selection, model ) {
+/**
+ * Returns a node that will be used to insert image with `model.insertContent`.
+ * @private
+ */
```
Comment (arkflpc) [inline packages/ckeditor5-image/src/imageutils.ts:280]:
```suggestion
 */
```

File: packages/ckeditor5-image/src/pictureediting.ts [hunk 16]
```diff
@@ -122,28 +125,29 @@ export default class PictureEditing extends Plugin {
...
-
-			if ( !sources ) {
-				return;
-			}
-
-			editor.model.change( writer => {
-				writer.setAttributes( {
-					sources
-				}, imageElement );
+		this.listenTo(
```
Comment (arkflpc) [inline packages/ckeditor5-image/src/pictureediting.ts:136]: It looks like deserving an event type.

File: packages/ckeditor5-ui/src/bindings/csstransitiondisablermixin.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		constructor( ...args: Array<any> ) {
+			super( ...args );
+
+			this.set( '_isCssTransitionsDisabled', false );
+			this.initializeMixin();
+		}
+
+		protected initializeMixin(): void {
```
Comment (arkflpc) [inline packages/ckeditor5-ui/src/bindings/csstransitiondisablermixin.ts:58]:
```suggestion
		protected initializeCssTransitionDisablerMixin(): void {
```

File: packages/ckeditor5-ui/src/bindings/injectcsstransitiondisabler.ts [hunk 18]
```diff
@@ -76,7 +78,6 @@ export default function injectCssTransitionDisabler( view: View ): asserts view
 }
 
 export type ViewWithCssTransitionDisabler = View & {
```
Comment (arkflpc) [inline packages/ckeditor5-ui/src/bindings/injectcsstransitiondisabler.ts:80]: I'd move it to \`[csstransitiondisablermixin.ts](https://github.com/ckeditor/ckeditor5/pull/13222/files#diff-8eb907dbb7cc5b2bcfb06c5d79810a5efb1169160189e80f5220b334e04e4fef)\`

File: packages/ckeditor5-image/src/imageupload/imageuploadprogress.ts [hunk 19]
```diff
@@ -9,71 +9,84 @@
...
+	public static get pluginName(): 'ImageUploadProgress' {
 		return 'ImageUploadProgress';
 	}
 
+	/**
+	 * The image placeholder that is displayed before real image data can be accessed.
+	 *
+	 * For the record, this image is a 1x1 px GIF with an aspect ratio set by CSS.
+	 */
+	protected placeholder: string;
```
[the remark this answers is not in the document]
Comment (mremiszewski) [reply packages/ckeditor5-image/src/imageupload/imageuploadprogress.ts:47]: I changed it to `private`

**Pull request merged**